### DISCUSSION TITLE
Strategy request: Target block coding at OpenSCAD capabilities

### DISCUSSION THREAD

**Discussion by goatchurchprime** [endlessm/godot-block-coding#378] [Ideas]
### What Happened

Godot [now has](https://github.com/godotengine/godot/pull/94321) the same CSG (Computational Solid Geometry) engine as [OpenSCAD](https://openscad.org/).  The reason OpenSCAD is so widely known and used is it solves a real problem of being able to quickly generate shapes for 3D printing with a gentle learning curve.  (However, speaking from experience, things get unmaintainably complex once you go much further.)

Block coding would be an perfect interface to this CSG engine, and the result would be more tangible since the operations are expressed in the Godot node-tree where they are easy to tinker with.

I've managed to implement a basic demo of the [Menger Sponge](https://manifoldcad.org/#Menger%20Sponge) using codeblocks (see below) without too many contrived functions at the top level called `EXF`.

The extra features I need to make this work properly are:

* Function calls with arguments
* Creating and parenting of new CSG nodes
* Setting and getting of CSG attributes (the library currently only handles Node2Ds as a special case)

To get round this I have called into `EXF` level functions `push_s()` and `pop_s()` to implement a stack due to the lack of local function variables, and `create_csg()` to make a new object, `push_child()` to parent the last object under the previous one, `duplicate_push_last()` a contrived compound function for this case, and `set_last_orientation()`. 

Once you can do function calls, the `EXF` level functions are a good place to stage any features in order to discover what is minimal.

I note that function names are text lines, while variables are drag and drop.  It's not clear how to reconcile this.  Although the text lines are more versatile, the drag and drop of names in bubbles is satisfying.  An in-between method is to drag a bubble onto a text line to fill that text line.

![Screenshot from 2024-12-05 10-30-50](https://github.com/user-attachments/assets/d222cdc8-c910-4d07-89ad-e1aa08652144)
![Screenshot from 2024-12-05 10-32-10](https://github.com/user-attachments/assets/833a637b-7581-4adb-b43e-71ec53f76695)

### How to Reproduce

Project is here [csgmanifoldcodeblockstest.zip](https://github.com/user-attachments/files/18022532/csgmanifoldcodeblockstest.zip)

### Block Coding Plugin Version

v0.7.1

### Godot Engine Version

v4.4.dev7 (not yet released)

### Operating System

Linux

### Logs or command-line output

```shell
N/A
```

### Anything else?

My long term goal is to find something as effective as [Protoflux](https://protoflux.wiki/#/protoflux/allNodes) in [Resonite](https://resonite.com/features.html] where I often see multiple people working socially on the same complex visual script in VR.  

Doing this for the generation of CSG objects that are to be 3D printed is an achievable goal where there is a realistic and good possibility of it becoming a notable success.

**Comment by manuq** [comment]
@goatchurchprime very impressive!

> The extra features I need to make this work properly are:
> 
> * Function calls with arguments

Yes, this needs to be implemented, tracked in #126 . This should be implemented as a new mini editor for the function definition, similar to the current one for defining variables:

![image](https://github.com/user-attachments/assets/1461f610-f4e4-4216-aa9a-8260e704e714)

> * Creating and parenting of new CSG nodes

Before adding more blocks it would be nice to have an "Advanced" filter. #282 . For preventing the interface to be too cluttered, and keep it for educational purposes.

In case you would like "3d printing" custom blocks to be shipped separately, see how custom blocks can be defined dynamically in scripts with `func setup_custom_blocks()`. Which is not great and can be improved: #329 .

> * Setting and getting of CSG attributes (the library currently only handles Node2Ds as a special case)

I'm happy to say that v0.8 (just released) comes with a feature to drag any property from the inspector to create a block for the attribute. See https://github.com/endlessm/godot-block-coding/releases/tag/v0.8.0

Still, you will probably need Vector3 support: #330

> I note that function names are text lines, while variables are drag and drop. It's not clear how to reconcile this. Although the text lines are more versatile, the drag and drop of names in bubbles is satisfying. An in-between method is to drag a bubble onto a text line to fill that text line.

Yes, this can certainly be improved.

**Comment by manuq** [comment]
@goatchurchprime I have moved this from issue to discussion. We have Vector3 support now. Did you had a chance to test the latest release of this plugin?

**Reply by goatchurchprime** [reply]
This project has almost got to the top of my pile several times, but there are too many other Godot coding things ongoing.  (eg I'm giving a 3 hour zoom lesson on Godot to some unreal game devs tomorrow, and I am not an experienced teacher.)  I've also been touring people around Resonite, which is very much the guiding light in terms of in-game collaborative coding.  If you have a good enough PCVR I'd really like to give you a tour of where this leads to in the extreme case, so the end goal is experienced and not simply speculation.  It's available now, but just not in low-power enough form.
